### PR TITLE
Support check expression inside split BIR function

### LINKED ISSUE #33838: Split BIR functions even when the return val is assigned value inside the split
**Description:**

When splitting the BIR functions into new methods, should also handle the case of making the split even though the return val is assigned value inside the split.

**Describe your task(s)**

**Related Issues (optional):**

#32895

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by Nadeeshan96**
$title
Fixes #33838

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java [hunk 1]
```diff
@@ -399,6 +430,38 @@ private void generateSplits(BIRFunction function, List<Split> possibleSplits,
...
+                                                 List<BIRBasicBlock> newBBList, BIROperand splitLastInsLhsOp) {
+        BIROperand isErrorResultOp = generateTempLocalVariable(function, symbolTable.booleanType);
+        BIRNonTerminator errorTestIns = new BIRNonTerminator.TypeTest(null, symbolTable.errorType,
+                isErrorResultOp, splitFuncCallResultOp);
+        errorTestIns.scope = lastInsScope;
+        currentBB.instructions.add(errorTestIns);
+        BIRBasicBlock trueBB = new BIRBasicBlock(new Name("bb" + ++newBBNum));
+        BIRBasicBlock falseBB = new BIRBasicBlock(new Name("bb" + ++newBBNum));
+        currentBB.terminator = new BIRTerminator.Branch(null, isErrorResultOp, trueBB, falseBB);
+        currentBB.terminator.scope = lastInsScope;
```
Comment (gabilang) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java:444]: Shall we use the constructor which included the bir scope as a parameter? And consider all other places too.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java [hunk 2]
```diff
@@ -134,16 +144,17 @@ private boolean hasLessInstructionCount(BIRFunction birFunction) {
             BIRTerminator bbTerminator = basicBlock.terminator;
             if (splitStarted) {
                 if (bbTerminator.lhsOp != null) {
-                    // as the varDcl is available inside the split no need to pass as function arg
-                    if (needToPassLhsVarDclAsArg(bbTerminator.lhsOp)) {
+                    if ((bbTerminator.lhsOp.variableDcl.kind == VarKind.LOCAL) ||
+                            (bbTerminator.lhsOp.variableDcl.kind == VarKind.RETURN)) {
+                        // if a local or the return var is assigned value in a BB terminator, no split is done
+                        splitStarted = false;
```
Comment (gabilang) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java:150]: Why did we consider VarKind.LOCAL here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java [hunk 3]
```diff
@@ -365,9 +380,20 @@ private void generateSplits(BIRFunction function, List<Split> possibleSplits,
...
-                    fromAttachedFunction);
+            BType newFuncReturnType = lastInstruction.lhsOp.variableDcl.type;
+            BIROperand splitLastInsLhsOp = new BIROperand(lastInstruction.lhsOp.variableDcl);
+            BIROperand splitFuncCallResultOp;
+            if (currSplit.returnValAssigned) {
+                splitFuncCallResultOp = generateTempLocalVariable(function, newFuncReturnType);
+                LinkedHashSet<BType> memberTypes = new LinkedHashSet<>(2);
+                memberTypes.add(newFuncReturnType);
+                memberTypes.add(symbolTable.errorType);
+                newFuncReturnType = new BUnionType(null, memberTypes, false, false);
```
Comment (gabilang) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java:391]: Shall we move this union type creation to a separate method?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java (hunk 2 above)
Comment (Nadeeshan96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java:150]: We cannot assign value for a local variable inside a new split function, as they should be passed as arguments. So if `VarKind.LOCAL` var is assigned value, no split is done. Here I think `bbTerminator.lhsOp.variableDcl.kind == VarKind.RETURN` won't happen most probably. But added that condition for the safety.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java (hunk 3 above)
Comment (Nadeeshan96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java:391]: Will look into it.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java (hunk 1 above)
Comment (Nadeeshan96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java:444]: For `BIRTerminator` I will use such constructor. For `BIRNonTerminator`s there is no constructor which included the bir scope as a parameter.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java [hunk 4]
```diff
@@ -642,6 +702,20 @@ private BIRBasicBlock generateSplitsInSameBB(BIRFunction function, int bbNum, Li
...
 
+    /**
+     * Generate a temporary function scope variable.
+     *
+     * @param func The BIR function to which the variable should be added
+     * @param variableType The type of the variable
+     * @return The generated operand for the variable declaration
+     */
+    private BIROperand generateTempLocalVariable(BIRFunction func, BType variableType) {
+        Name variableName = new Name("$split$tempVar$" + splitTempVarNum++);
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java:713]:
```suggestion
        Name variableName = new Name("$split$tempVar$_" + splitTempVarNum++);
```

**Comment by Nadeeshan96** [comment]
Closing this PR as these changes are already there, because stage-slbeta6 branch has been merged to the master branch.

**Pull request closed without merging**